patches/entropy-0.4.patch

```diff
@@ -0,0 +1,694 @@
++*.aux
++*.hp
++*.eventlog
++.stack-work/
++cabal.project.local
++.HTF/
++
++
++# End of https://www.gitignore.io/api/haskell
+diff --git a/System/Entropy.hs b/System/Entropy.hs
+index db809f0..96489b2 100644
+--- a/System/Entropy.hs
++++ b/System/Entropy.hs
+@@ -22,21 +22,13 @@ module System.Entropy
+           hGetEntropy,
+           closeHandle
+         ) where
+-#if defined(isWindows)
+-import System.EntropyWindows
+-#else
+-#ifdef XEN
+-import System.EntropyXen
+-#else
+-import System.EntropyNix
+-#endif
+-#endif
++import System.EntropyJVM
+ 
+ import qualified Data.ByteString as B
+ import Control.Exception (bracket)
+ 
+ -- |Get a specific number of bytes of cryptographically
+--- secure random data using the *system-specific* sources. 
++-- secure random data using the *system-specific* sources.
+ -- (As of 0.4.  Verions <0.4 mixed system and hardware sources)
+ --
+ -- The returned random value is considered cryptographically secure but not true entropy.
+diff --git a/System/EntropyJVM.hs b/System/EntropyJVM.hs
+new file mode 100644
+index 0000000..d60b279
+--- /dev/null
++++ b/System/EntropyJVM.hs
+@@ -0,0 +1,73 @@
++{-# LANGUAGE CPP, ForeignFunctionInterface, BangPatterns, ScopedTypeVariables, MagicHash #-}
++{-|
++ Maintainer: Thomas.DuBuisson@gmail.com
++ Stability: beta
++ Portability: portable
++
++ Obtain entropy from system sources or x86 RDRAND when available.
++
++-}
++
++module System.EntropyJVM
++        ( CryptHandle
++        , openHandle
++        , hGetEntropy
++        , closeHandle
++        , hardwareRandom
++        ) where
++
++import Control.Monad (liftM, when)
++import Data.ByteString as B
++import System.IO.Error (mkIOError, eofErrorType, ioeSetErrorString)
++import Data.Bits (xor)
++
++import Foreign.Ptr
++import Foreign.C.Types
++
++import Java
++import Data.ByteString.Internal as B
++
++foreign import java unsafe "@static Utils.get_rand_bytes"
++  java_get_rand_bytes :: Ptr CUChar -> CSize -> IO CInt
++
++-- import System.Posix (openFd, closeFd, fdReadBuf, OpenMode(..), defaultFileFlags, Fd)
++
++-- |Handle for manual resource management
++data CryptHandle
++    = CH SecureRandom
++
++data SecureRandom = SecureRandom @java.security.SecureRandom
++  deriving Class
++
++foreign import java unsafe "@new" newSecureRandom :: IO SecureRandom
++
++
++-- | Get random values from the hardward RNG or return Nothing if no
++-- supported hardware RNG is available.
++--
++-- Supported hardware:
++--      * RDRAND
++--      * Patches welcome
++hardwareRandom :: Int -> IO (Maybe B.ByteString)
++hardwareRandom n = do
++  random <- newSecureRandom
++  Just <$> hGetEntropy (CH random) n
++
++-- |Open a `CryptHandle`
++openHandle :: IO CryptHandle
++openHandle = do
++  CH `fmap` newSecureRandom
++
++-- |Close the `CryptHandle`
++closeHandle :: CryptHandle -> IO ()
++closeHandle (CH _) = return ()
++
++-- |Read random data from a `CryptHandle`
++hGetEntropy :: CryptHandle -> Int -> IO B.ByteString
++hGetEntropy (CH _) n =
```

Comment:
You create a `SecureRandom` instance but it doesn't look like you re-use it - we can save some allocations if it is. Maybe pass it into `java_get_rand_bytes` as an extra argument?